README.rst
=========================
Music Full Tag Downloader
=========================


.. image:: https://img.shields.io/pypi/v/musicftdl.svg
        :target: https://pypi.python.org/pypi/musicftdl

.. image:: https://img.shields.io/travis/lonsty/musicftdl.svg
        :target: https://travis-ci.com/lonsty/musicftdl

.. image:: https://readthedocs.org/projects/musicftdl/badge/?version=latest
        :target: https://musicftdl.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




A CLI tool to download music of Jay Chou and other singers with full song tags.


* Free software: MIT license
* Documentation: https://musicftdl.readthedocs.io.


Features
--------

* **Search**: Search songs by keywords.
* **Download**: Download song(s) by singer/album/song mid or keyword.
* **Tag**: After downloading, tags are automatically added to the song (eg artist, album, sum of tracks, album cover image, release date), this feature is only available for `MP3` format songs.


HISTORY.rst
=======
History
=======

0.1.0 (2020-05-11)
-------------------

New feature:

* Download songs of selected types of albums (Studio Album, EP Single, Live Album, Others).
* Print error messages to help users troubleshoot.

Fixed bug:

* Fixed an issue where empty files were created when songs could not be downloaded.

0.1.0a (2020-05-10)
-------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for musicftdl."""


tests/test_musicftdl.py
#!/usr/bin/env python

"""Tests for `musicftdl` package."""


import unittest

from click.testing import CliRunner

from musicftdl import cli, musicftdl


class TestMusicftdl(unittest.TestCase):
    """Tests for `musicftdl` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_000_something(self):
        """Test something."""

    def test_command_line_interface(self):
        """Test the CLI."""
        runner = CliRunner()
        result = runner.invoke(cli.cli)
        assert result.exit_code == 0
        assert '--help  Show this message and exit.' in result.output
        help_result = runner.invoke(cli.cli, ['--help'])
        assert help_result.exit_code == 0
        assert '--help  Show this message and exit.' in help_result.output


musicftdl/cli.py
"""Console script for musicftdl."""
import sys
from typing import List

import click

from musicftdl.models import DownloadArgs
from musicftdl.musicftdl import download as dl
from musicftdl.musicftdl import (get_album_songs, get_singer_albums,
                                 get_song_info, get_song_url)
from musicftdl.musicftdl import search as search_kw
from musicftdl.utils import cut_str_to_multi_line, print_table


@click.group()
def cli():
    """A CLI tool to download music of Jay Chou and other singers with full song tags."""
    return 0


@cli.command()
# @click.option('-t', '--type', default='0', show_default=True,
#               type=click.Choice(['0', '2', '7', '8', '9', '12']),
#               help='0: song\n2: song list\n7: lyrics\n8: album\n9: singer\n12: mv')
@click.option('-P', '--page', default=1, show_default=True, help='Page No.')
@click.option('-S', '--page-size', default=20, show_default=True, help='Page size.')
@click.argument('keywords', nargs=-1)
def search(keywords, page, page_size):
    """Search songs by KEYWORDS."""
    try:
        result = search_kw(' '.join(keywords), page, page_size)
    except Exception as e:
        print(e)
        sys.exit(1)

    if result:
        headers = result[0].dict(exclude={'str_media_mid'}).keys()
        rows = [['\n'.join(c) if isinstance(c, List) else c  # noqa
                 for c in item.dict(exclude={'str_media_mid'}).values()]  # noqa
                for item in result]  # noqa

        print_table(headers, rows)


@cli.command()
@click.option('-P', '--page', default=1, show_default=True, help='Page No.')
@click.option('-S', '--page-size', default=50, show_default=True, help='Page size.')
@click.argument('mid')
def list(mid, page, page_size):
    """List albums/songs of the given SINGER/ALBUM MID."""
    try:
        result = get_singer_albums(mid, page, page_size)
    except Exception as e:
        print(e)
        sys.exit(1)

    if result:
        # for item in result:
        #     item.album_name = cut_str_to_multi_line(item.album_name, 12)
        headers = result[0].dict(exclude={'album_cover_url', 'album_cover_content',
                                          'songs', 'album_index', 'language'}).keys()
        rows = [['\n'.join(c) if isinstance(c, List) else c
                 for c in item.dict(exclude={'album_cover_url', 'album_cover_content',
                                             'songs', 'album_index', 'language'}).values()]
                for item in result]

        print_table(headers, rows)
        return

    try:
        result = get_album_songs(mid)
    except Exception as e:
        print(e)
        sys.exit(1)

    if result:
        headers = result[0].dict(exclude={'url', 'str_media_mid', 'genre'}).keys()
        rows = [['\n'.join(c) if isinstance(c, List) else c
                 for c in item.dict(exclude={'url', 'str_media_mid', 'genre'}).values()]
                for item in result]

        print_table(headers, rows)


@cli.command()
@click.argument('song-mid')
def show(song_mid):
    """Show information and play url of the given SONG MID."""
    try:
        song_info = get_song_info(song_mid)
    except Exception as e:
        print(e)
        sys.exit(1)

    if not song_info.song_mid:
        print(f'Song <{song_mid}> not found!')
        sys.exit(1)

    intro = song_info.introduction
    if intro:
        song_info.introduction = cut_str_to_multi_line(intro, 50)
    headers = ['Category', 'Description']
    rows = [['\n'.join(c) if isinstance(c, List) else c for c in item]
            for item in song_info.dict(exclude={'url', 'str_media_mid', 'album_cover_url',
                                                'album_cover_content', 'album_singers_mid', 'album_singers_name'}).items()]
    print_table(headers, rows)
    try:
        print(get_song_url(song_mid))
    except Exception as e:
        print(e)


@cli.command()
@click.option('-s', '--singer', is_flag=True, help='Download songs by SINGER_MID.')
@click.option('-a', '--album', is_flag=True, help='Download songs by ALBUM_MID.')
@click.option('-k', '--keywords', is_flag=True, help='Download song by searching keywords.')
@click.option('-o', '--overwrite', is_flag=True, help='Overwrite exist files.')
@click.option('-d', '--destination', default='.', help='Destination to save songs.')
@click.option('-n', '--name-style', default='3', show_default=True,
              type=click.Choice(['1', '2', '3']),
              help='Style of filename. [ 1: SONG.ext | 2: SINGER - SONG.ext | 3: SINGER - ALBUM - SONG.ext')
@click.option('-t', '--album_types', default='SELO', show_default=True,
              help='Download songs of selected types of albums. [ S: Studio Album | '
                   'E: EP Single | L: Live Album | O: Others ]')
@click.option('-c/-C', '--classified/--no-classified', 'classified', default=True,
              show_default=True, help='Store in folders classify by singers and albums.')
@click.option('-f', '--format', default='128', show_default=True,
              type=click.Choice(['128', '320', 'm4a', 'flac', 'ape']), help='Song format.')
@click.option('-P', '--page', default=1, show_default=True, help='Page No.')
@click.option('-S', '--page-size', default=50, show_default=True, help='Page size.')
# @click.option('--proxy', default=None, help='Set a HTTP/HTTPS proxy, as format USERNAME:PASSWORD@IP:PORT.')
@click.argument('resource')
def download(**kwargs):
    """Download songs by SINGER/ALBUM MID or KEYWORDS."""
    args = DownloadArgs(**kwargs)
    try:
        dl(args)
    except Exception as e:
        print(e)


if __name__ == "__main__":
    sys.exit(cli())  # pragma: no cover


musicftdl/utils.py
# @Author: allen
# @Date: May 06 14:36 2020
import math
import os
import random
import time
from datetime import timedelta
from functools import wraps

from prettytable import PrettyTable


def print_table(title, items):
    table = PrettyTable(title, border=True, hrules=1, header_style='upper', align='m')
    for item in items: table.add_row(item)
    print(table)


def retry(exceptions, tries=3, delay=1, backoff=2, logger=None):
    """
    Retry calling the decorated function using an exponential backoff.
    Args:
        exceptions: The exception to check. may be a tuple of
            exceptions to check.
        tries: Number of times to try (not retry) before giving up.
        delay: Initial delay between retries in seconds.
        backoff: Backoff multiplier (e.g. value of 2 will double the delay
            each retry).
        logger: Logger to use. If None, print.
    """

    def deco_retry(f):

        @wraps(f)
        def f_retry(*args, **kwargs):
            mtries, mdelay = tries, delay or random.uniform(0.5, 1.5)
            while mtries > 1:
                try:
                    return f(*args, **kwargs)
                except exceptions as e:
                    if logger:
                        logger.warning('{}, Retrying in {} seconds...'.format(e, mdelay))
                    time.sleep(mdelay)
                    mtries -= 1
                    mdelay *= backoff
            return f(*args, **kwargs)

        return f_retry

    return deco_retry


def mkdirs_if_not_exist(dir):
    if not os.path.isdir(dir):
        try:
            os.makedirs(dir)
        except FileExistsError:
            pass


def convert_to_safe_filename(filename):
    return "".join([c for c in filename if c not in r'\/:*?"<>|']).strip()


# def convert_seconds_to_dtstr(sec: int) -> str:
#     p = str(timedelta(seconds=sec)).split(':')
#     return ':'.join(p if int(p[0]) > 0 else p[1:])


def convert_seconds_to_dtstr(sec: int) -> str:
    dt_str = str(timedelta(seconds=sec))
    return dt_str[-5:] if sec < 3600 else dt_str


def cut_str_to_multi_line(s, l=50):
    """

    :param s: string
    :param l: length
    :return:
    """
    return '\n'.join([s[i * l:(i + 1) * l] for i in range(math.ceil(len(s) / l))])


musicftdl/models.py
# @Author: allen
# @Date: May 05 15:56 2020
import os
from typing import List

from dateutil import parser
from pydantic import BaseModel, validator

from musicftdl.utils import (convert_seconds_to_dtstr,
                             convert_to_safe_filename, mkdirs_if_not_exist)


class API(BaseModel):
    search: str = 'http://mapi.lonsty.me/search?key={}&pageNo={}&pageSize={}'
    singer_album: str = 'http://mapi.lonsty.me/singer/album?singermid={}&pageNo={}&pageSize={}'
    album: str = 'http://mapi.lonsty.me/album?albummid={}'
    album_songs: str = 'http://mapi.lonsty.me/album/songs?albummid={}'
    song_info: str = 'http://mapi.lonsty.me/song?songmid={}'
    song_url: str = 'http://mapi.lonsty.me/song/url?id={}&type={}'


class Consts(BaseModel):
    api: API = API()
    headers: dict = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
                                   '(KHTML, like Gecko) Chrome/81.0.4044.113 Safari/537.36'}
    proxies = {
        'http': 'http://183.129.244.16:33238',
        'https': 'https://183.129.244.16:33238'
    }
    timeout: tuple = (15, 30)


class SearchResult(BaseModel):
    singers_mid: list= None
    singers_name: list = None
    album_mid: str = None
    album_name: str = None
    song_mid: str = None
    song_name: str = None
    duration: int = None
    str_media_mid: str = None

    @validator('duration')
    def convert_int_to_str(cls, v):
        return convert_seconds_to_dtstr(v)

    @property
    def singer_mid(self):
        return self.singers_mid[0] if self.singers_mid else None

    @property
    def singer_name(self):
        return '&'.join(self.singers_name) if self.singers_name else None


class Song(BaseModel):
    singers_mid: list= None
    singers_name: list = None
    song_mid: str = None
    song_name: str = None
    song_index: int = None
    duration: int = None
    genre: int = None
    url: str = None
    str_media_mid: str = None

    @validator('duration')
    def convert_int_to_str(cls, v):
        return convert_seconds_to_dtstr(v)

    @property
    def singer_mid(self):
        return self.singers_mid[0] if self.singers_mid else None

    @property
    def singer_name(self):
        return '&'.join(self.singers_name) if self.singers_name else None


class Album(BaseModel):
    singers_mid: list = None
    singers_name: list = None
    album_mid: str = None
    album_name: str = None
    album_type: str = None
    album_index: int = None
    language: str = None
    song_count: int = None
    company: str = None
    publish_time: str = None
    album_cover_url: str = None
    album_cover_content: bytes = None
    songs: List[Song] = None

    @property
    def singer_mid(self):
        return self.singers_mid[0] if self.singers_mid else None

    @property
    def singer_name(self):
        return '&'.join(self.singers_name) if self.singers_name else None

    @property
    def publish_date(self):
        return parser.parse(self.publish_time)

    @property
    def album_cover_bg_url(self):
        if self.album_mid:
            return f'https://y.gtimg.cn/music/photo_new/T002R800x800M000{self.album_mid}.jpg?max_age=2592000'

    @property
    def singer_pic_bg_url(self):
        if self.singer_mid:
            return f'https://y.gtimg.cn/music/photo_new/T001R800x800M000{self.singer_mid}.jpg?max_age=2592000'


class Singer(BaseModel):
    singer_mid: str = None
    singer_name: str = None
    albums: List[Album] = None


class SongInfo(BaseModel):
    song_mid: str = None
    song_name: str = None
    duration: int = None
    singers_mid: list= None
    singers_name: list = None
    album_mid: str = None
    album_name: str = None
    album_singers_mid: list = None
    album_singers_name: list = None
    song_index: int = None
    company: str = None
    introduction: str = None
    language: str = None
    genre: int = None
    publish_time: str = None
    url: str = None
    album_cover_url: str = None
    album_cover_content: bytes = None
    str_media_mid: str = None

    @validator('duration')
    def convert_int_to_str(cls, v):
        return convert_seconds_to_dtstr(v)


    @property
    def singer_name(self):
        return '&'.join(self.singers_name) if self.singers_name else None

    @property
    def album_singer_name(self):
        return '&'.join(self.album_singers_name) if self.album_singers_name else self.singer_name

    @property
    def publish_date(self):
        return parser.parse(self.publish_time)

    @property
    def album_cover_bg_url(self):
        if self.album_mid:
            return f'https://y.gtimg.cn/music/photo_new/T002R800x800M000{self.album_mid}.jpg?max_age=2592000'


class DownloadArgs(BaseModel):
    resource: str = None
    singer: bool = False
    album: bool = False
    keywords: bool = False
    overwrite: bool = False
    destination: str = None
    name_style: int = None
    album_types: str = 'SELO'
    classified: bool = True
    format: str = None
    page: int = None
    page_size: int = None
    proxy: str = None

    @property
    def extension(self):
        return 'mp3' if self.format in ['128', '320'] else self.format

    def format_name(self, song: SongInfo) -> str:
        if self.name_style == 3:
            basename = f'{song.singer_name} - {song.album_name} - {song.song_name}.{self.extension}'
        elif self.name_style == 2:
            basename = f'{song.singer_name} - {song.song_name}.{self.extension}'
        else:
            basename = f'{song.song_name}.{self.extension}'
        return convert_to_safe_filename(basename)

    def filename(self, song: SongInfo) -> str:
        basename = self.format_name(song)
        paths = [self.destination,
                 convert_to_safe_filename(song.album_singer_name),
                 convert_to_safe_filename(song.album_name)]
        folder = os.path.join(*paths) if self.classified else paths[0]
        mkdirs_if_not_exist(folder)
        return os.path.join(folder, basename)

    def filter_albums(self, albums: List[Album]) -> List[Album]:
        types_mp = {'S': '录音室专辑', 'E': 'EP单曲', 'L': '现场专辑', 'O': '其他'}
        types = [types_mp.get(k) for k in list(self.album_types.upper())]
        return [album for album in albums if album.album_type in types or '其他' in types]

    @property
    def retag(self):
        return True if self.format in ['128', '320'] else False

    @property
    def proxies(self):
        proxies = None
        if self.proxy:
            proxies = {
                'http': f'http://{self.proxy}',
                'https': f'https://{self.proxy}'
            }
        return proxies


musicftdl/id3_genres.py
# @Author: allen
# @Date: May 08 18:01 2020

class Genre:
    genres = {
        0: 'Blues',
        1: 'Classic rock',
        2: 'Country',
        3: 'Dance',
        4: 'Disco',
        5: 'Funk',
        6: 'Grunge',
        7: 'Hip-Hop',
        8: 'Jazz',
        9: 'Metal',
        10: 'New Age',
        11: 'Oldies',
        12: 'Other',
        13: 'Pop',
        14: 'Rhythm and Blues',
        15: 'Rap',
        16: 'Reggae',
        17: 'Rock',
        18: 'Techno',
        19: 'Industrial',
        20: 'Alternative',
        21: 'Ska',
        22: 'Death metal',
        23: 'Pranks',
        24: 'Soundtrack',
        25: 'Euro-Techno',
        26: 'Ambient',
        27: 'Trip-Hop',
        28: 'Vocal',
        29: 'Jazz & Funk',
        30: 'Fusion',
        31: 'Trance',
        32: 'Classical',
        33: 'Instrumental',
        34: 'Acid',
        35: 'House',
        36: 'Game',
        37: 'Sound clip',
        38: 'Gospel',
        39: 'Noise',
        40: 'Alternative Rock',
        41: 'Bass',
        42: 'Soul',
        43: 'Punk',
        44: 'Space',
        45: 'Meditative',
        46: 'Instrumental Pop',
        47: 'Instrumental Rock',
        48: 'Ethnic',
        49: 'Gothic',
        50: 'Darkwave',
        51: 'Techno-Industrial',
        52: 'Electronic',
        53: 'Pop-Folk',
        54: 'Eurodance',
        55: 'Dream',
        56: 'Southern Rock',
        57: 'Comedy',
        58: 'Cult',
        59: 'Gangsta',
        60: 'Top 40',
        61: 'Christian Rap',
        62: 'Pop/Funk',
        63: 'Jungle',
        64: 'Native US',
        65: 'Cabaret',
        66: 'New Wave',
        67: 'Psychedelic',
        68: 'Rave',
        69: 'Show tunes',
        70: 'Trailer',
        71: 'Lo-Fi',
        72: 'Tribal',
        73: 'Acid Punk',
        74: 'Acid Jazz',
        75: 'Polka',
        76: 'Retro',
        77: 'Musical',
        78: 'Rock ’n’ Roll',
        79: 'Hard Rock',
        80: 'Folk',
        81: 'Folk-Rock',
        82: 'National Folk',
        83: 'Swing',
        84: 'Fast Fusion',
        85: 'Bebop',
        86: 'Latin',
        87: 'Revival',
        88: 'Celtic',
        89: 'Bluegrass',
        90: 'Avantgarde',
        91: 'Gothic Rock',
        92: 'Progressive Rock',
        93: 'Psychedelic Rock',
        94: 'Symphonic Rock',
        95: 'Slow rock',
        96: 'Big Band',
        97: 'Chorus',
        98: 'Easy Listening',
        99: 'Acoustic',
        100: 'Humour',
        101: 'Speech',
        102: 'Chanson',
        103: 'Opera',
        104: 'Chamber music',
        105: 'Sonata',
        106: 'Symphony',
        107: 'Booty bass',
        108: 'Primus',
        109: 'Porn groove',
        110: 'Satire',
        111: 'Slow jam',
        112: 'Club',
        113: 'Tango',
        114: 'Samba',
        115: 'Folklore',
        116: 'Ballad',
        117: 'Power ballad',
        118: 'Rhythmic Soul',
        119: 'Freestyle',
        120: 'Duet',
        121: 'Punk Rock',
        122: 'Drum solo',
        123: 'A cappella',
        124: 'Euro-House',
        125: 'Dance Hall',
        126: 'Goa',
        127: 'Drum & Bass',
        128: 'Club-House',
        129: 'Hardcore Techno',
        130: 'Terror',
        131: 'Indie',
        132: 'BritPop',
        133: 'Negerpunk',
        134: 'Polsk Punk',
        135: 'Beat',
        136: 'Christian Gangsta Rap',
        137: 'Heavy Metal',
        138: 'Black Metal',
        139: 'Crossover',
        140: 'Contemporary Christian',
        141: 'Christian rock',
        142: 'Merengue',
        143: 'Salsa',
        144: 'Thrash Metal',
        145: 'Anime',
        146: 'Jpop',
        147: 'Synthpop',
        148: 'Abstract',
        149: 'Art Rock',
        150: 'Baroque',
        151: 'Bhangra',
        152: 'Big beat',
        153: 'Breakbeat',
        154: 'Chillout',
        155: 'Downtempo',
        156: 'Dub',
        157: 'EBM',
        158: 'Eclectic',
        159: 'Electro',
        160: 'Electroclash',
        161: 'Emo',
        162: 'Experimental',
        163: 'Garage',
        164: 'Global',
        165: 'IDM',
        166: 'Illbient',
        167: 'Industro-Goth',
        168: 'Jam Band',
        169: 'Krautrock',
        170: 'Leftfield',
        171: 'Lounge',
        172: 'Math Rock',
        173: 'New Romantic',
        174: 'Nu-Breakz',
        175: 'Post-Punk',
        176: 'Post-Rock',
        177: 'Psytrance',
        178: 'Shoegaze',
        179: 'Space Rock',
        180: 'Trop Rock',
        181: 'World Music',
        182: 'Neoclassical',
        183: 'Audiobook',
        184: 'Audio theatre',
        185: 'Neue Deutsche Welle',
        186: 'Podcast',
        187: 'Indie-Rock',
        188: 'G-Funk',
        189: 'Dubstep',
        190: 'Garage Rock',
        191: 'Psybient'
    }

    inverted_genres = {v: k for k, v in genres.items()}

    @classmethod
    def get_genre(cls, num: int) -> str:
        return cls.genres.get(num, 'Unknown')

    @classmethod
    def get_num(cls, genre: str) -> int:
        return cls.inverted_genres.get(genre, 12)


musicftdl/musicftdl.py
"""Main module."""
import os
import threading
from concurrent.futures import ThreadPoolExecutor, as_completed, wait
from typing import List

import eyed3
from requests import Response, Session, exceptions

from musicftdl.id3_genres import Genre
from musicftdl.models import (Album, Consts, DownloadArgs, SearchResult,
                              Singer, Song, SongInfo)
from musicftdl.utils import retry

consts = Consts()
thread_local = threading.local()
eyed3.log.setLevel("ERROR")


def _get_session() -> Session:
    """
    Get a instance of requests.Session
    :return: requests.Session
    """
    if not hasattr(thread_local, "session"):
        thread_local.session = Session()
    return thread_local.session


class DataNotFoundError(Exception):
    pass


@retry(Exception)
def session_request(url: str, method: str='GET') -> Response:
    session = _get_session()

    with session.request(method, url, headers=consts.headers, timeout=consts.timeout) as resp:
        resp.raise_for_status()
        return resp


def get_json_data(resp):
    try:
        res = resp.json()
    except Exception as e:
        print(e)
        return {}
    else:
        if res.get('result') != 100:
            raise DataNotFoundError(res.get('errMsg'))
        data = res.get('data')
        if not data:
            raise DataNotFoundError('Data is empty')
        return data


def search(key: str, page: int=1, page_size: int=20) -> List[SearchResult]:
    data = get_json_data(session_request(consts.api.search.format(key, page, page_size)))

    result = []
    for item in data.get('list', []):
        result.append(SearchResult(
            singers_mid=[it.get('mid') for it in item.get('singer', [{}])],
            singers_name=[it.get('name') for it in item.get('singer', [{}])],
            album_name=item.get('albumname'),
            album_mid=item.get('albummid'),
            song_name=item.get('songname'),
            song_mid=item.get('songmid'),
            duration=item.get('interval'),
            str_media_mid=item.get('strMediaMid', '')
        ))
    return result


def get_singer_albums(singer_mid: str, page: int=1, page_size: int=50) -> List[Album]:
    data = get_json_data(session_request(consts.api.singer_album.format(singer_mid, page, page_size)))

    result = []
    for idx, item in enumerate(reversed(data.get('list', []))):
        result.append(Album(
            album_mid=item.get('album_mid'),
            album_name=item.get('album_name'),
            album_type=item.get('albumtype'),
            album_index=idx + 1,
            language=item.get('lan'),
            song_count=item.get('latest_song', {}).get('song_count'),
            company=item.get('company', {}).get('company_name'),
            publish_time=item.get('pub_time'),
            singers_mid=[it.get('singer_mid') for it in item.get('singers', [{}])],
            singers_name=[it.get('singer_name') for it in item.get('singers', [{}])]
        ))
    return result


def get_album(album_mid: str) -> Album:
    data = get_json_data(session_request(consts.api.album.format(album_mid)))

    album = Album(
        album_mid=data.get('mid'),
        album_name=data.get('name'),
        album_type=None,
        album_index=None,
        company=data.get('company'),
        publish_time=data.get('publishTime'),
        singers_mid=[it.get('mid') for it in data.get('ar', [{}])],
        singers_name=[it.get('name') for it in data.get('ar', [{}])],
        album_cover_url=data.get('picUrl')
    )
    return album


def get_album_cover(album_mid: str) -> str:
    data = get_json_data(session_request(consts.api.album.format(album_mid)))

    return 'http:' + data.get('picUrl')


def get_album_songs(album_mid: str) -> List[Song]:
    data = get_json_data(session_request(consts.api.album_songs.format(album_mid)))
    result = []

    for idx, item in enumerate(data.get('list', [])):
        result.append(Song(
            song_mid=item.get('mid'),
            song_name=item.get('name'),
            song_index=item.get('index_album'),
            singers_mid=[it.get('mid') for it in item.get('singer', [{}])],
            singers_name=[it.get('name') for it in item.get('singer', [{}])],
            duration=item.get('interval'),
            genre=item.get('genre'),
            str_media_mid=item.get('file', {}).get('media_mid')
        ))
    return result


def get_song_url(song_mid: str, format: str= '128') -> str:
    return get_json_data(session_request(consts.api.song_url.format(song_mid, format)))


def get_song_info(song_mid: str) -> SongInfo:
    data = get_json_data(session_request(consts.api.song_info.format(song_mid)))
    song_info = SongInfo(
        song_mid=data.get('track_info', {}).get('mid'),
        song_name=data.get('track_info', {}).get('name'),
        duration=data.get('track_info', {}).get('interval'),
        singers_mid=[it.get('mid') for it in data.get('track_info', {}).get('singer', [{}])],
        singers_name=[it.get('name') for it in data.get('track_info', {}).get('singer', [{}])],
        album_mid=data.get('track_info', {}).get('album', {}).get('mid'),
        album_name=data.get('track_info', {}).get('album', {}).get('name'),
        song_index=data.get('track_info', {}).get('index_album'),
        company=data.get('info', {}).get('company', {}).get('content', [{}])[0].get('value'),
        genre=Genre.get_num(data.get('info', {}).get('genre', {}).get('content', [{}])[0].get('value')),
        introduction=data.get('info', {}).get('intro', {}).get('content', [{}])[0].get('value'),
        language=data.get('info', {}).get('lan', {}).get('content', [{}])[0].get('value'),
        publish_time=data.get('info', {}).get('pub_time', {}).get('content', [{}])[0].get('value')
    )
    return song_info


def download_music(url: str, filename: str, overwrite: bool=False):
    if os.path.isfile(filename) \
        and not overwrite \
        and os.path.getsize(filename) > 0:
        print(f'{filename} [Skipped]')
        return False

    try:
        resp = session_request(url)
    except Exception as e:
        print(e)
    else:
        if resp is not None:
            with open(filename, 'wb') as f:
                for chunk in resp.iter_content(8192):
                    f.write(chunk)
            print(filename)
            return True


def add_tags(filename: str, song_info: SongInfo):
    try:
        audiofile = eyed3.load(filename)

        audiofile.tag.title = song_info.song_name
        audiofile.tag.artist = song_info.singer_name
        # audiofile.tag.genre = song_info.genre
        audiofile.tag.album = song_info.album_name
        audiofile.tag.album_artist = song_info.album_singer_name
        # audiofile.tag.album_type = song.genre
        audiofile.tag.track_num = (song_info.song_index, 0)
        # audiofile.tag.disc_num = (None, None)
        audiofile.tag.publisher = song_info.company
        audiofile.tag.copyright = song_info.company
        audiofile.tag.recording_date = song_info.publish_date
        audiofile.tag.release_date = song_info.publish_time
        # audiofile.tag.best_release_date = song.publish_date
        audiofile.tag.images.set(3, song_info.album_cover_content, 'image/jpeg', song_info.album_name + '.JPG')
        audiofile.tag.save(encoding='utf-8')
    except Exception:
        pass


def download_with_tags(filename: str, song_info: SongInfo, overwrite: bool=False, retag: bool=True):
    downloaded = download_music(song_info.url, filename, overwrite)
    if downloaded and retag:
        add_tags(filename, song_info)


def fetch_album_chore(album: Album, args: DownloadArgs) -> Album:
    with ThreadPoolExecutor(max_workers=20) as pool:
        album.songs = get_album_songs(album.album_mid)
        song_url_futures = {pool.submit(get_song_url, song.song_mid, args.format): song
                            for song in album.songs}

        cover_resp = session_request(album.album_cover_bg_url)
        album.album_cover_content = cover_resp.content if cover_resp.status_code == 200 else b''

        download_futures = {}
        for future in as_completed(song_url_futures):
            try:
                result = future.result()
            except Exception as e:
                song = song_url_futures[future]
                print(f'{e}, {song.singer_name} - {album.album_name} - '
                      f'[{song.song_mid} {args.format}] {song.song_name}')
            else:
                song = song_url_futures[future]
                song.url = result
                song_info = SongInfo(
                    song_mid=song.song_mid,
                    song_name=song.song_name,
                    singers_mid=song.singers_mid,
                    singers_name=song.singers_name,
                    album_mid=album.album_mid,
                    album_name=album.album_name,
                    album_singers_mid=album.singers_mid,
                    album_singers_name=album.singers_name,
                    album_cover_url=album.album_cover_bg_url,
                    album_cover_content=album.album_cover_content,
                    song_index=song.song_index,
                    company=album.company,
                    genre=song.genre,
                    introduction=None,
                    language=album.language,
                    publish_time=album.publish_time,
                    url=result,
                    str_media_mid=song.str_media_mid
                )
                download_futures[pool.submit(download_with_tags, args.filename(song_info),
                                             song_info, args.overwrite, args.retag)] = song_info
        wait(download_futures)
    return album


def _download_by_song_mid(song_mid, args):
    song_info = get_song_info(song_mid)
    assert song_info.song_mid, f'Song <{song_mid}> not found!'
    song_info.url = get_song_url(song_info.song_mid, args.format)
    song_info.album_cover_content = session_request(song_info.album_cover_bg_url).content
    download_with_tags(args.filename(song_info), song_info, args.overwrite, args.retag)


def download(args: DownloadArgs):
    if not any([args.singer, args.album, args.keywords]):
        return _download_by_song_mid(args.resource, args)

    if args.singer:
        albums = get_singer_albums(args.resource, page=args.page, page_size=args.page_size)
        assert albums, f'Singer <{args.resource}> not found!'
        singer = Singer(singer_mid=args.resource, singer_name=albums[0].singer_name, albums=albums)
    elif args.album:
        album = get_album(args.resource)
        assert album.album_mid, f'Album <{args.resource}> not found!'
        singer = Singer(singer_mid=album.singer_mid, singer_name=album.singer_name, albums=[album])
    else:
        result = search(args.resource, page=1, page_size=1)
        assert result, f'Resource <{args.resource}> not found!'
        return _download_by_song_mid(result[0].song_mid, args)

    for album in args.filter_albums(singer.albums):
        fetch_album_chore(album, args)
